### PR TITLE
Additional annotation parsing

### PR DESCRIPTION
**Pull request by reicheratwork**
Signed-off-by: Martijn Reicher <martijn.reicher@adlinktech.com>

### CODE REVIEW

**Comment by k0ekk0ek** [comment]
Thanks @reicheratwork! Better to have a member in the corresponding nodes to indicate if something's optional so that each backend can use it. Also, would be nice to verify `@optional` can only be assigned to the correct types (probably structs and unions only?).

**Comment by reicheratwork** [comment]
> Thanks @reicheratwork! Better to have a member in the corresponding nodes to indicate if something's optional so that each backend can use it. Also, would be nice to verify `@optional` can only be assigned to the correct types (probably structs and unions only?).

Can't @optional not be assigned to basic types as well?
If i recall correctly, it is an annotation to members, right?

**Comment by reicheratwork** [comment]
@k0ekk0ek in my most recent push, I added an idl_boolean_t is_optional to the idl_member_t class and created a utility function checking for this, so that the different backend implementations will not have to create their own

File: src/idl/src/annotation.c [hunk 1]
```diff
@@ -133,6 +133,31 @@ annotate_autoid(
...
+
+  assert(annotation_appl);
+  if (annotation_appl->parameters
+   && annotation_appl->parameters->const_expr
+   && idl_mask(annotation_appl->parameters->const_expr) != (IDL_LITERAL | IDL_BOOL)) {
+        idl_error(pstate, idl_location(annotation_appl),
+          "@optional cannot have non-boolean values assigned");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
```
Comment (dennis-adlink) [inline src/idl/src/annotation.c:157]: Shouldn't `((idl_member_t *)node)->is_optional` be set at this point?

File: src/idl/src/annotation.c [hunk 2]
```diff
@@ -261,6 +286,29 @@ annotate_key(
...
+  if (annotation_appl->parameters
+   && annotation_appl->parameters->const_expr) {
+        idl_error(pstate, idl_location(annotation_appl),
+          "@default can only be assigned to values of the following types: ");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }*/
+
+  return IDL_RETCODE_OK;
+}
+
```
Comment (dennis-adlink) [inline src/idl/src/annotation.c:311]: Maybe better to move the @default implementation to a different PR (including the changes below for @default in the annotations array)? Or fix the implementation in this PR (and change the PR title)

File: src/idl/src/parser.y [hunk 3]
```diff
@@ -812,8 +812,11 @@ case_label:
   ;
 
 element_spec:
-    type_spec declarator
-      { TRY(idl_create_case(pstate, LOC(@1.first, @2.last), $1, $2, &$$)); }
+    annotations type_spec declarator
+      { TRY(idl_create_case(pstate, LOC(@1.first, @3.last), $2, $3, &$$));
+        TRY_EXCEPT(idl_annotate(pstate, $$, $1), free($$)); }
+  | type_spec declarator
+      { TRY(idl_create_case(pstate, LOC(@1.first, @2.last), $1, $2, &$$));}
```
Comment (dennis-adlink) [inline src/idl/src/parser.y:819]: @optional cannot be applied to union members

File: src/idl/src/tree.c [hunk 4]
```diff
@@ -772,6 +772,59 @@ uint32_t idl_bound(const void *node)
...
+bool idl_is_optional(const void *node)
+{
+  if (idl_is_declarator(node)) {
+    const idl_node_t *parent = idl_parent(node);
+    if (idl_is_member(parent))
+      return idl_is_optional(parent);
+  } else {
+    idl_annotation_appl_t *annotation_appl = NULL;
+    IDL_FOREACH(annotation_appl, ((const idl_node_t*)node)->annotations) {
+      if (0 != strcmp(idl_identifier(annotation_appl->annotation), "optional"))
```
Comment (dennis-adlink) [inline src/idl/src/tree.c:784]: Although we don't have a coding style guide and slightly different styles are used through the code base, I'd prefer a consistent coding style in a single code file, so `strmp(..) != 0` or just `strcmp(...)`

File: src/idl/src/tree.c [hunk 5]
```diff
@@ -772,6 +772,59 @@ uint32_t idl_bound(const void *node)
...
+      } else {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+const idl_literal_t *idl_has_default(const void *node)
+{
```
Comment (dennis-adlink) [inline src/idl/src/tree.c:801]: Move to a separate PR?

Comment (dennis-adlink) [reply src/idl/src/tree.c:801]: With the function name `idl_has_default` I would expect a bool that indicates if the node has a default (which is what the function `idl_is_default` below does). Maybe a better name would be `idl_get_default_value` or something similar?

File: src/idl/src/tree.c [hunk 6]
```diff
@@ -772,6 +772,59 @@ uint32_t idl_bound(const void *node)
...
+  return NULL;
+}
+
+bool idl_is_default(const void *node)
+{
+  const idl_literal_t *lit = idl_has_default(node);
+
+  if (lit
+   && idl_mask(lit) == (IDL_LITERAL | IDL_ANY))
+    return lit->value.bln;
```
Comment (dennis-adlink) [inline src/idl/src/tree.c:823]: Does this also work if an explicit default value `0` is applied to a member (`@default(0)`), this function should return true in that case?

File: src/idl/tests/annotation.c [hunk 7]
```diff
@@ -67,6 +67,124 @@ CU_Test(idl_annotation, id_member)
...
+  assert(u);
+  IDL_FOREACH(uc, u->cases) {
+    CU_ASSERT_PTR_NOT_NULL_FATAL(uc);
+    CU_ASSERT_FATAL(idl_is_case(uc));
+    CU_ASSERT_EQUAL(idl_is_optional(uc), n++ > 0);
+  }
+
+  idl_delete_pstate(pstate);
+
+  const char *errstr[] = {"@optional stuct s {\n"
```
Comment (dennis-adlink) [inline src/idl/tests/annotation.c:136]: Typo `stuct` ?

File: src/idl/tests/annotation.c [hunk 8]
```diff
@@ -67,6 +67,124 @@ CU_Test(idl_annotation, id_member)
...
+                          "  char c;\n"
+                          "};\n",
+                          "enum e {\n"
+                          "  e_0,\n"
+                          "  @optional e_1,\n"
+                          "  e_2,\n"
+                          "  e_3\n"
+                          "};\n"};
+
```
Comment (dennis-adlink) [inline src/idl/tests/annotation.c:145]: Maybe add a test for invalid arguments, like `@optional(1)`, `@optional('true')` ?

File: src/idl/tests/annotation.c [hunk 9]
```diff
@@ -67,6 +67,124 @@ CU_Test(idl_annotation, id_member)
...
+  idl_declarator_t *d;
+  idl_union_t *u;
+  idl_case_t *uc;
+  const char structstr[] = "struct s {\n"
+                           "  char c_none;\n"
+                           "  @optional(false) char c_false;\n"
+                           "  @optional char c_default;\n"
+                           "  @optional(true) char c_true;\n"
+                           "  @optional char c_1_default, c_2_default;\n"
+                           "};\n";
```
Comment (dennis-adlink) [inline src/idl/tests/annotation.c:89]: Add test some cases for other member types, e.g. aggregated types, sequences, arrays?

**Review by dennis-adlink — COMMENTED** [review]
I've reviewed these changes and found a few issues that need a fix, mainly the incomplete @default implementation and using the is_optional field instead of iterating through the annotations in `idl_is_optional`. And some nit-picking on the consistency of the coding style.

**Comment by reicheratwork** [comment]
> I've reviewed these changes and found a few issues that need a fix, mainly the incomplete @default implementation and using the is_optional field instead of iterating through the annotations in `idl_is_optional`. And some nit-picking on the consistency of the coding style.

Thanks for taking the time to look over this PR! I will take your comments into account, though I will probably rebase the current changes onto #923 when it has been merged into master. After which some of the new features' implementation may be quite different.

**Comment by k0ekk0ek** [comment]
The `IDL_ANNOTATABLE` changes are in the master branch.

**Comment by reicheratwork** [comment]
> The `IDL_ANNOTATABLE` changes are in the master branch.

Super!
I will rebase this onto the latest changes and then rework it into the IDL_ANNOTATABLE syntax

File: src/idl/include/idl/tree.h [hunk 10]
```diff
@@ -299,6 +299,7 @@ struct idl_member {
   idl_declarator_t *declarators;
   /* metadata */
   idl_boolean_t key;
+  idl_boolean_t is_optional;
```
[the remark this answers is not in the document]
Comment (reicheratwork) [reply src/idl/include/idl/tree.h]: You are correct, in the newest iteration this is fixed.
However, some member names cannot be chosen as they are reserved keywords (such as default), so there I may need to use little workarounds.

File: src/idl/src/parser.y (hunk 3 above)
Comment (reicheratwork) [reply src/idl/src/parser.y]: You are correct, this has been removed.

File: src/idl/tests/annotation.c (hunk 7 above)
Comment (reicheratwork) [reply src/idl/tests/annotation.c]: You are correct.

File: src/idl/tests/annotation.c (hunk 8 above)
Comment (reicheratwork) [reply src/idl/tests/annotation.c]: The number of test cases will be expanded a lot in the next iteration.

File: src/idl/tests/annotation.c (hunk 9 above)
Comment (reicheratwork) [reply src/idl/tests/annotation.c]: Will do.

File: src/idl/src/annotation.c [hunk 11]
```diff
@@ -133,6 +133,31 @@ annotate_autoid(
...
 }
 
+static idl_retcode_t
+annotate_optional(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  if (!idl_is_member(node)
+   && !idl_is_case(node)) {
```
[the remark this answers is not in the document]
Comment (reicheratwork) [reply src/idl/src/annotation.c]: This  is corrected in the new iteration.

File: src/idl/src/annotation.c (hunk 1 above)
Comment (reicheratwork) [reply src/idl/src/annotation.c:158]: This is no longer relevant, after rebasing I have changed to using the IDL_ANNOTATABLE macro

File: src/idl/src/annotation.c [hunk 12]
```diff
@@ -306,6 +341,85 @@ set_nested(
...
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+  mem_spec = ((idl_member_t*)node)->type_spec;
+
+  value = annotation_appl->parameters->const_expr;
+  if (!value || !idl_is_literal(value)) {
+    idl_error(pstate, idl_location(annotation_appl),
+      "@default requires a value");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
```
Comment (k0ekk0ek) [inline src/idl/src/annotation.c:368]: This can be an `assert`. The parser checks the syntax and since there's no default, it should throw an error if there's no value specified. Probably good to add a small test for that (if there's none yet). If the test fails, we (or I) should fix the bug there.

File: src/idl/src/annotation.c [hunk 13]
```diff
@@ -306,6 +341,85 @@ set_nested(
...
+    default:
+    {
+      idl_mask_t value_type = (idl_mask(value) ^ IDL_LITERAL);
+      if (idl_mask(mem_spec) != value_type) {
+        idl_error(pstate, idl_location(annotation_appl),
+          "value type of @default does not match member");
+        return IDL_RETCODE_SEMANTIC_ERROR;
+      }
+    }
+  }
```
Comment (k0ekk0ek) [inline src/idl/src/annotation.c:395]: This too, should have been done by the parser already.

File: src/idl/src/annotation.c [hunk 14]
```diff
@@ -129,6 +129,41 @@ annotate_autoid(
...
+  if (annotation_appl->parameters) {
+    const_expr = annotation_appl->parameters->const_expr;
+    assert(const_expr);
+    if (idl_mask(const_expr) != (IDL_BOOL | IDL_LITERAL)) {
+      idl_error(pstate, idl_location(annotation_appl),
+        "@optional cannot have non-boolean values assigned");
+      return IDL_RETCODE_SEMANTIC_ERROR;
+    } else {
+      value = ((const idl_literal_t*)const_expr)->value.bln;
+    }
```
Comment (k0ekk0ek) [inline src/idl/src/annotation.c:158]: This is probably checked by the parser already. Like in one of my earlier comments, let's add a couple of small tests and fix the parser if it doesn't behave correctly already.

File: src/idl/src/annotation.c [hunk 15]
```diff
@@ -306,6 +341,85 @@ set_nested(
...
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  assert(pstate);
+
+  if (!idl_is_enumerator(node)) {
+        idl_error(pstate, idl_location(annotation_appl),
+          "@default_literal can only be assigned to member nodes");
+        return IDL_RETCODE_SEMANTIC_ERROR;
```
Comment (k0ekk0ek) [inline src/idl/src/annotation.c:414]: Indentation.

File: src/idl/include/idl/tree.h [hunk 16]
```diff
@@ -378,6 +380,7 @@ struct idl_enumerator {
      mapped to a native data type capable of representing a maximally-sized
      enumeration */
   IDL_ANNOTATABLE(uint32_t) value;
+  IDL_ANNOTATABLE(bool) is_default;
```
Comment (k0ekk0ek) [inline src/idl/include/idl/tree.h:383]: In closing the `enum` declaration we should check if there's only one default. Plus, I'm thinking it may actually make more sense to have a `default_enumerator` member in the `enum` too/instead?

Comment (k0ekk0ek) [reply src/idl/include/idl/tree.h:383]: The `default_enumerator` could then point to the first enumerator is not explicit default is provided? (Probably works like that, but I'm not a 100% sure)

Comment (k0ekk0ek) [reply src/idl/include/idl/tree.h:383]: Maybe introduce `IDL_DEFAULT_ENUMERATOR` that equals `IDL_ENUMERATOR | 1u`. The (implicit) default case for unions works like that as well. Don't know if it makes sense, but it might be worth looking into that.

File: src/idl/src/annotation.c (hunk 12 above)
Comment (reicheratwork) [reply src/idl/src/annotation.c:368]: You're correct that this should not really be checked here.
I will check whether the parser catches this, or if this requires a fix someplace else

File: src/idl/src/annotation.c (hunk 13 above)
Comment (reicheratwork) [reply src/idl/src/annotation.c:395]: I think that there might be a misunderstanding here, this is checking whether the type in the literal (the parameter field of the `@default` annotation) is compatible with that of the type it is being annotated to (so it should reject something like this: `@default('a') long l;`

File: src/idl/include/idl/tree.h (hunk 16 above)
Comment (reicheratwork) [reply src/idl/include/idl/tree.h:383]: My attempt here was to modify the existing interface as little as possible, but your approach should result in less duplicated functionality, as the different backends will not have to implement their own checks

Comment (reicheratwork) [reply src/idl/include/idl/tree.h:383]: I will look into the default member approach

File: src/idl/src/annotation.c (hunk 14 above)
Comment (reicheratwork) [reply src/idl/src/annotation.c:158]: You're correct that this should not really be checked here.
I will check whether the parser catches this, or if this requires a fix someplace else

File: src/idl/src/annotation.c (hunk 12 above)
Comment (reicheratwork) [reply src/idl/src/annotation.c:368]: This is correctly being checked by the parser at a previous step, and therefore I will replace this check with a simpler assert

File: src/idl/src/annotation.c (hunk 14 above)
Comment (reicheratwork) [reply src/idl/src/annotation.c:158]: This is correctly being checked by the parser at a previous step, and therefore I will replace this check with a simpler assert

File: src/idl/src/annotation.c (hunk 15 above)
Comment (reicheratwork) [reply src/idl/src/annotation.c:414]: Will be fixed

File: src/idl/tests/annotation.c [hunk 17]
```diff
@@ -45,6 +45,196 @@ parse_string(uint32_t flags, const char *str, idl_pstate_t **pstatep)
...
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional(true) char c; };"                 },
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional(true) char c_1, c_2, c_3; };"     },
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional(true) char c_1, c_2, c_3; };"     },
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional sequence<double> s_d; };"         },
+    {IDL_RETCODE_OK,                  true,   "typedef sequence<long> seq_long;\n"
+                                              "struct s { @optional seq_long s_l_a[15]; };"           },
+    {IDL_RETCODE_SEMANTIC_ERROR,      false,  "@optional struct s { char c; };"                       },
+    {IDL_RETCODE_ILLEGAL_EXPRESSION,  false,  "struct s { @optional(\"true\") char c; };"             },
+    {IDL_RETCODE_ILLEGAL_EXPRESSION,  false,  "struct s { @optional(123) char c; };"                  },
+    {IDL_RETCODE_ILLEGAL_EXPRESSION,  false,  "struct s { @optional(0.1) char c; };"                  },
```
Comment (k0ekk0ek) [inline src/idl/tests/annotation.c:98]: I think it'd be nicer to cover these in more generic annotation tests(?)

File: src/idl/tests/annotation.c [hunk 18]
```diff
@@ -45,6 +45,196 @@ parse_string(uint32_t flags, const char *str, idl_pstate_t **pstatep)
...
+    {IDL_RETCODE_SEMANTIC_ERROR,      false,  "enum e {  e_0, @optional e_1};"                        },
+    {IDL_RETCODE_SYNTAX_ERROR,        false,  "union u switch (long) { case 0: @optional double d; };"}
```
Comment (k0ekk0ek) [inline src/idl/tests/annotation.c:100]: Same for this one because we're actually testing the syntax rather than something `@optional` specific.

File: src/idl/tests/annotation.c (hunk 17 above)
Comment (reicheratwork) [reply src/idl/tests/annotation.c:98]: My reasoning behind putting the tests for the correct parameter type in the annotation is that this is something that is unique to the @optional annotation, and in my opinion, having a whole forest of different checks of matching the parameter type with the annotation type would be less descriptive.

File: src/idl/tests/annotation.c (hunk 18 above)
Comment (reicheratwork) [reply src/idl/tests/annotation.c:100]: Same reasoning as previous comment.

File: src/idl/include/idl/tree.h (hunk 16 above)
Comment (reicheratwork) [reply src/idl/include/idl/tree.h:383]: Changed to a default_enumerator pointer in the idl_enum class

File: src/idl/tests/annotation.c [hunk 19]
```diff
@@ -45,6 +45,210 @@ parse_string(uint32_t flags, const char *str, idl_pstate_t **pstatep)
...
+    {"struct s { @default(123) string str; };",              IDL_RETCODE_SEMANTIC_ERROR,  false, IDL_NULL,    NULL}, //parameter type mismatch (int vs string)
+    {"struct s { @default(\"false\") boolean b; };",         IDL_RETCODE_SEMANTIC_ERROR,  false, IDL_NULL,    NULL}, //parameter type mismatch (string vs bool)
+    {"struct s { @default(123) boolean b; };",               IDL_RETCODE_SEMANTIC_ERROR,  false, IDL_NULL,    NULL}, //parameter type mismatch (int vs bool)
+    {"union u (switch long) {\n"
+     "case 1: @default(123) short s;\n"
+     "case 2: string s; };",                                 IDL_RETCODE_SYNTAX_ERROR,    false, IDL_NULL,    NULL}, //unions do not allow defaults on values
+    /* skipping this test as idl_create_annotation_appl leaks memory if idl_resolve cannot resolve the scoped name (https://github.com/eclipse-cyclonedds/cyclonedds/issues/950)
+      {"@default(e_0) enum e { e_0, e_1, e_2, e_3 };",         IDL_RETCODE_SEMANTIC_ERROR,  false, IDL_NULL,    NULL}  //setting default on enums is done through @default_literal
+    */
```
Comment (k0ekk0ek) [inline src/idl/tests/annotation.c:187]: Better enable this and indicate the test is known to fail using the [WILL_FAIL property](https://cmake.org/cmake/help/latest/prop_test/WILL_FAIL.html). Then, when the fix for #950 is in, the test will pass (causing the build to fail). The advantage being that we're actively notified when a change is required.

File: src/idl/tests/annotation.c [hunk 20]
```diff
@@ -45,6 +45,210 @@ parse_string(uint32_t flags, const char *str, idl_pstate_t **pstatep)
...
+          CU_ASSERT_EQUAL(idl_mask(en), IDL_ENUMERATOR);
+        }
+      }
+    }
+
+    idl_delete_pstate(pstate);
+  }
+}
+
+CU_Test(idl_annotation, enum_default)
```
Comment (k0ekk0ek) [inline src/idl/tests/annotation.c:236]: If you can find the time, please rename the testcase to `default_literal`. That way it's directly obvious in the logs/terminal what this test is meant to test (not that it's super hard to guess, but still).

File: src/idl/tests/annotation.c (hunk 19 above)
Comment (reicheratwork) [reply src/idl/tests/annotation.c:187]: The issue is not here that the test fails, it is that there is a memory leak that is triggered by this test, and that the builds in Azure that check for memory leaks will fail if this test is done.

File: src/idl/tests/annotation.c (hunk 20 above)
Comment (reicheratwork) [reply src/idl/tests/annotation.c:236]: Will do.

File: src/idl/tests/annotation.c (hunk 19 above)
Comment (k0ekk0ek) [reply src/idl/tests/annotation.c:187]: I know, but CTest will mark it as a failure. Since `WILL_FAIL` is a CTest property, it'll get us the desired behavior right? If you prefer you may set the test to disabled. It's just that commenting it out only ensures it's never looked after again. At least, there's tons of commented out tests (or test ideas) for libidl that would've caught a bunch of things but they were never looked at or thought of again... I prefer to have some kind of notification that something is still off, at least that way no one forgets.

**Review by k0ekk0ek — APPROVED** [review]

**Comment by k0ekk0ek** [comment]
Just a couple of warnings in the tests, if those are solved, I'll happily merge

**Comment by reicheratwork** [comment]
> Just a couple of warnings in the tests, if those are solved, I'll happily merge

That one escaped my notice, I will fix it immediately.
